src/DotNetWorker.Core/FunctionsApplication.cs

```diff
@@ -17,28 +20,52 @@ internal partial class FunctionsApplication : IFunctionsApplication
         private readonly ConcurrentDictionary<string, FunctionDefinition> _functionMap = new ConcurrentDictionary<string, FunctionDefinition>();
         private readonly FunctionExecutionDelegate _functionExecutionDelegate;
         private readonly IFunctionContextFactory _functionContextFactory;
+        private readonly FunctionInvocationManager _functionInvocationManager;
         private readonly IOptions<WorkerOptions> _workerOptions;
         private readonly ILogger<FunctionsApplication> _logger;
         private readonly IWorkerDiagnostics _diagnostics;
 
-        public FunctionsApplication(FunctionExecutionDelegate functionExecutionDelegate, IFunctionContextFactory functionContextFactory,
-             IOptions<WorkerOptions> workerOptions, ILogger<FunctionsApplication> logger, IWorkerDiagnostics diagnostics)
+        public FunctionsApplication(
+            FunctionExecutionDelegate functionExecutionDelegate,
+            IFunctionContextFactory functionContextFactory,
+            FunctionInvocationManager functionInvocationManager,
+            IOptions<WorkerOptions> workerOptions,
+            ILogger<FunctionsApplication> logger,
+            IWorkerDiagnostics diagnostics)
         {
             _functionExecutionDelegate = functionExecutionDelegate ?? throw new ArgumentNullException(nameof(functionExecutionDelegate));
             _functionContextFactory = functionContextFactory ?? throw new ArgumentNullException(nameof(functionContextFactory));
+            _functionInvocationManager = functionInvocationManager ?? throw new ArgumentNullException(nameof(functionInvocationManager));
             _workerOptions = workerOptions ?? throw new ArgumentNullException(nameof(workerOptions));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
             _diagnostics = diagnostics ?? throw new ArgumentNullException(nameof(diagnostics));
         }
 
         public FunctionContext CreateContext(IInvocationFeatures features)
         {
+            CancellationTokenSource? cancellationTokenSource = null;
             var invocation = features.Get<FunctionInvocation>() ?? throw new InvalidOperationException($"The {nameof(FunctionInvocation)} feature is required.");
 
             var functionDefinition = _functionMap[invocation.FunctionId];
             features.Set<FunctionDefinition>(functionDefinition);
 
-            return _functionContextFactory.Create(features);
+            if (functionDefinition.Parameters.Where(p => p.Type == typeof(CancellationToken)).Any())
```

Comment:
I don't love that we have to loop through the Parameters every time. Is there a way we can cache whether or not we need to create a CTS for this function? It's not going to change, and this happens with every invocation.

Reply:
Great point, will look into caching that information

Reply:
Using lazy init in the function definition to cache if the function is using a cancellation token or not - does this work or is there a better way to do this?

Reply:
Spoke with Fabio and he mentioned that we should always be creating a cancellation token so this becomes irrelevant